### PR TITLE
Added transfert function editor

### PR DESCRIPTION
**Pull request by chevtche**

### CODE REVIEW

File: apps/livreGUI/CMakeLists.txt [hunk 1]
```diff
@@ -15,38 +16,33 @@ set(LIVREGUI_HEADERS
...
-  editor/graphs/TransferFunctionGraph.cpp)
+  editor/ColorMapWidget.cpp
+  editor/GradientRenderer.cpp
+  editor/HoverPoints.cpp
+  editor/TFWidget.cpp
+  editor/TransferFunctionEditor.cpp)
 
 set(LIVREGUI_RESOURCES resources/resources.qrc)
-set(LIVREGUI_UI_FORMS editor/gui/TransferFunctionEditor.ui)
+set(LIVREGUI_UI_FORMS editor/TransferFunctionEditor.ui)
```
Comment (bilgili) [inline apps/livreGUI/CMakeLists.txt]: into a gui folder

File: apps/livreGUI/editor/ColorMapWidget.cpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <livreGUI/editor/ColorMapWidget.h>
+#include <livreGUI/editor/HoverPoints.h>
+
+namespace livre
+{
+
+ColorMapWidget::ColorMapWidget( const SHADE_TYPE type, QWidget* colorMapParentWidget )
```
Comment (bilgili) [inline apps/livreGUI/editor/ColorMapWidget.cpp]: rename the enum according to our standard. enum HelloWorld

File: apps/livreGUI/editor/ColorMapWidget.cpp [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+{
+    // Add checkers background for the alpha widget.
+    if ( _shadeType == ARGB_SHADE )
+    {
+        _createCheckersBackground();
+    }
+    else
+        setAttribute( Qt::WA_NoBackground );
+
+    QPolygonF points;
```
Comment (bilgili) [inline apps/livreGUI/editor/ColorMapWidget.cpp]: remove it

File: apps/livreGUI/editor/ColorMapWidget.cpp [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+QPolygonF ColorMapWidget::getPoints() const
+{
+    return _hoverPoints->points();
+}
+
+uint ColorMapWidget::getColorAtPoint( const int xPosition )
+{
+    _generateShade();
+
+    const QPolygonF hoverPoints = _hoverPoints->points();
```
Comment (bilgili) [inline apps/livreGUI/editor/ColorMapWidget.cpp]: const QPolygonF&

File: apps/livreGUI/editor/ColorMapWidget.cpp [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    for ( int i = 1; i < hoverPoints.size(); ++i )
```
Comment (bilgili) [inline apps/livreGUI/editor/ColorMapWidget.cpp]: size_t

File: apps/livreGUI/editor/ColorMapWidget.cpp [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    {
+        if ( hoverPoints.at(i - 1).x() <= xPosition && hoverPoints.at(i).x() >= xPosition )
```
Comment (bilgili) [inline apps/livreGUI/editor/ColorMapWidget.cpp]: if(

File: apps/livreGUI/editor/ColorMapWidget.cpp [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        {
+            QLineF pointsConnection( hoverPoints.at( i-1 ), hoverPoints.at( i ) );
```
Comment (bilgili) [inline apps/livreGUI/editor/ColorMapWidget.cpp]: const

File: apps/livreGUI/editor/ColorMapWidget.cpp [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+            return _shade.pixel( qRound( qMin( pointsConnection.x2(),
+                                            ( qreal(_shade.width() - 1 )))),
+                                 qRound( qMin( pointsConnection.y2(),
+                                              qreal( _shade.height() - 1 ))));
+        }
+    }
+    return (uint) 0;
+}
+
+const std::vector<uint8_t> ColorMapWidget::getCurve()
```
Comment (bilgili) [inline apps/livreGUI/editor/ColorMapWidget.cpp]: use typdef , there is in livre core for Uint8.

Comment (bilgili) [inline apps/livreGUI/editor/ColorMapWidget.cpp]: no need for const

File: apps/livreGUI/editor/ColorMapWidget.cpp [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+{
+    std::vector<uint8_t> curve;
+
+    float delta = 256.0f / float( width() );
+    for( int i = 0; i < _hoverPoints->points().size(); ++i )
+    {
+        QLineF currentLine( _hoverPoints->points().at( i ), _hoverPoints->points().at( i + 1 ));
+
+        uint32_t beginPos = currentLine.p1().x() * delta + 0.5f;
```
Comment (bilgili) [inline apps/livreGUI/editor/ColorMapWidget.cpp]: consts

File: apps/livreGUI/editor/ColorMapWidget.cpp [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+
+void ColorMapWidget::setGradientStops( const QGradientStops& stops )
+{
+    if ( _shadeType == ARGB_SHADE )
+    {
+        _gradient = QLinearGradient( 0.0f, 0.0f, width(), 0.0f );
+
+        for ( int i = 0; i < stops.size(); ++i )
+        {
+            QColor color = stops.at( i ).second;
```
Comment (bilgili) [inline apps/livreGUI/editor/ColorMapWidget.cpp]: const ref

File: apps/livreGUI/editor/ColorMapWidget.h [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    /**
+     * paintEvent
+     */
+    void paintEvent( QPaintEvent* ) Q_DECL_OVERRIDE;
+
+    /**
+     * sizeHint
+     * @return Return the size.
+     */
+    QSize sizeHint() const Q_DECL_OVERRIDE;
```
Comment (bilgili) [inline apps/livreGUI/editor/ColorMapWidget.h]: final or override instead.

File: apps/livreGUI/editor/ColorMapWidget.h [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+     * @param xPosition Position of the mouse cursor on the x-axis.
+     * @return An uint containing RGBA 8-bits values.
+     */
+    uint getColorAtPoint(const int xPosition );
+
+    /**
+     * Return curve's values.
+     * @return A vector of 8-bits uints with the curve values.
+     */
+    const std::vector<uint8_t> getCurve();
```
Comment (bilgili) [inline apps/livreGUI/editor/ColorMapWidget.h]: typedef

File: apps/livreGUI/editor/GradientRenderer.h [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    /**
+     * Set the gradient stops.
+     * @param stops The stop points for gradient calculation.
+     */
+    void setGradientStops( const QGradientStops& stops );
+
+    /**
+     * paintEvent
+     */
+    void paintEvent( QPaintEvent* ) Q_DECL_OVERRIDE;
```
Comment (bilgili) [inline apps/livreGUI/editor/GradientRenderer.h]: ditto

File: apps/livreGUI/editor/HoverPoints.cpp [hunk 14]
```diff
@@ -0,0 +1,546 @@
...
+
+    widget->installEventFilter( this );
+    widget->setAttribute( Qt::WA_AcceptTouchEvents );
+
+    _connectionType = CURVE_CONNECTION;
+    _sortType = NO_SORT;
+    _pointShape = shape;
+    _pointPen = QPen( QColor( 255, 255, 255, 191 ), 1 );
+    _connectionPen = QPen( QColor( 255, 255, 255, 127 ), 2 );
+    _pointBrush = QBrush( QColor( 191, 191, 191, 127 ));
```
Comment (bilgili) [inline apps/livreGUI/editor/HoverPoints.cpp]: initialization list

File: apps/livreGUI/editor/HoverPoints.cpp [hunk 15]
```diff
@@ -0,0 +1,546 @@
...
+        {
+        case QEvent::MouseButtonPress:
+        {
+            if( !_fingerPointMapping.isEmpty())
+                return true;
+
+            QMouseEvent* mouseEvent = (QMouseEvent*)hoverEvent;
+            QPointF clickPosition = mouseEvent->pos();
+            int index = -1;
+            for( int i = 0; i < _tfPoints.size(); ++i )
```
Comment (bilgili) [inline apps/livreGUI/editor/HoverPoints.cpp]: size_t

File: apps/livreGUI/editor/TransferFunctionEditor.cpp [hunk 16]
```diff
@@ -21,114 +18,221 @@
...
+    , _publisher( 0 )
 {
-    ui_.setupUi( this );
-    setWindowFlags( (windowFlags( ) ^ Qt::Dialog) | Qt::Window);
+    ui->setupUi( this );
+
+    // Create the different colormap widgets for RGBA and add them to the UI
+    _redWidget = new ColorMapWidget( ColorMapWidget::RED_SHADE, this );
+    _greenWidget = new ColorMapWidget( ColorMapWidget::GREEN_SHADE, this );
+    _blueWidget = new ColorMapWidget( ColorMapWidget::BLUE_SHADE, this );
```
Comment (bilgili) [inline apps/livreGUI/editor/TransferFunctionEditor.cpp]: init list

File: apps/livreGUI/editor/TransferFunctionEditor.cpp [hunk 17]
```diff
@@ -21,114 +18,221 @@
...
+    std::sort( bluePoints.begin(), bluePoints.end(), xLessThan );
+    std::sort( alphaPoints.begin(), alphaPoints.end(), xLessThan );
+
+    allPoints = redPoints;
+    allPoints += greenPoints;
+    allPoints += bluePoints;
+    allPoints += alphaPoints;
+    std::sort( allPoints.begin(), allPoints.end(), xLessThan );
+
+    for ( int i = 0; i < allPoints.size(); ++i )
```
Comment (bilgili) [inline apps/livreGUI/editor/TransferFunctionEditor.cpp]: size_t

File: apps/livreGUI/CMakeLists.txt (hunk 1 above)
Comment (chevtche) [inline apps/livreGUI/CMakeLists.txt]: Done

File: apps/livreGUI/editor/ColorMapWidget.cpp (hunk 2 above)
Comment (chevtche) [inline apps/livreGUI/editor/ColorMapWidget.cpp]: Done

File: apps/livreGUI/editor/ColorMapWidget.cpp (hunk 4 above)
Comment (chevtche) [inline apps/livreGUI/editor/ColorMapWidget.cpp]: Done

File: apps/livreGUI/editor/ColorMapWidget.cpp (hunk 5 above)
Comment (chevtche) [inline apps/livreGUI/editor/ColorMapWidget.cpp]: Done

File: apps/livreGUI/editor/ColorMapWidget.cpp (hunk 6 above)
Comment (chevtche) [inline apps/livreGUI/editor/ColorMapWidget.cpp]: Done

File: apps/livreGUI/editor/ColorMapWidget.cpp (hunk 8 above)
Comment (chevtche) [inline apps/livreGUI/editor/ColorMapWidget.cpp]: Done

File: apps/livreGUI/editor/ColorMapWidget.cpp (hunk 9 above)
Comment (chevtche) [inline apps/livreGUI/editor/ColorMapWidget.cpp]: Done

File: apps/livreGUI/editor/ColorMapWidget.cpp (hunk 10 above)
Comment (chevtche) [inline apps/livreGUI/editor/ColorMapWidget.cpp]: Done

File: apps/livreGUI/editor/ColorMapWidget.h (hunk 11 above)
Comment (chevtche) [inline apps/livreGUI/editor/ColorMapWidget.h]: done

File: apps/livreGUI/editor/ColorMapWidget.h (hunk 12 above)
Comment (chevtche) [inline apps/livreGUI/editor/ColorMapWidget.h]: Done

File: apps/livreGUI/editor/GradientRenderer.h (hunk 13 above)
Comment (chevtche) [inline apps/livreGUI/editor/GradientRenderer.h]: done

File: apps/livreGUI/editor/HoverPoints.cpp (hunk 14 above)
Comment (chevtche) [inline apps/livreGUI/editor/HoverPoints.cpp]: Done

File: apps/livreGUI/editor/HoverPoints.cpp (hunk 15 above)
Comment (chevtche) [inline apps/livreGUI/editor/HoverPoints.cpp]: Done

File: apps/livreGUI/editor/TransferFunctionEditor.cpp (hunk 16 above)
Comment (chevtche) [inline apps/livreGUI/editor/TransferFunctionEditor.cpp]: Done

File: apps/livreGUI/editor/TransferFunctionEditor.cpp (hunk 17 above)
Comment (chevtche) [inline apps/livreGUI/editor/TransferFunctionEditor.cpp]: Done

File: apps/livreGUI/editor/HoverPoints.cpp [hunk 18]
```diff
@@ -0,0 +1,544 @@
...
+        qPainter.setPen( _connectionPen );
+
+        // Connect the points using smoothed curves.
+        if( _connectionType == CURVE_CONNECTION )
+        {
+            QPainterPath path;
+            path.moveTo( _tfPoints.at( 0 ));
+            for( int32_t i = 1; i < _tfPoints.size(); ++i )
+            {
+                QPointF p1 = _tfPoints.at( i - 1 );
```
Comment (bilgili) [inline apps/livreGUI/editor/HoverPoints.cpp]: where are my consts ?

File: apps/livreGUI/editor/HoverPoints.h [hunk 19]
```diff
@@ -0,0 +1,265 @@
...
+
+    /**
+     * Draw the control point.
+     */
+    void paintPoints();
+
+    /**
+     * Get the bounding rectangle of the point.
+     */
+    inline QRectF getBoundingRect() const;
```
Comment (bilgili) [inline apps/livreGUI/editor/HoverPoints.h]: remove inline

File: apps/livreGUI/editor/HoverPoints.h [hunk 20]
```diff
@@ -0,0 +1,265 @@
...
+
+public:
+
+    /**
+     * Fires pointsChanged() when one of the point changed.
+     */
+    void firePointChange();
+
+private:
+    inline QRectF _pointBoundingRectangle( int i ) const;
```
Comment (bilgili) [inline apps/livreGUI/editor/HoverPoints.h]: ditto

File: apps/livreGUI/editor/HoverPoints.h [hunk 21]
```diff
@@ -0,0 +1,265 @@
...
+    QVector<uint> _locks;
+    bool _editable;
+    bool _enabled;
+    QHash<int, int> _fingerPointMapping;
+};
+
+inline QRectF HoverPoints::_pointBoundingRectangle( const int i ) const
+{
+    QPointF p = _tfPoints.at(i);
+    qreal w = _pointSize.width();
```
Comment (bilgili) [inline apps/livreGUI/editor/HoverPoints.h]: consts ? I did not mark it all, check them throughly.

File: apps/livreGUI/editor/HoverPoints.cpp (hunk 18 above)
Comment (chevtche) [inline apps/livreGUI/editor/HoverPoints.cpp]: Done

File: apps/livreGUI/editor/HoverPoints.h (hunk 19 above)
Comment (chevtche) [inline apps/livreGUI/editor/HoverPoints.h]: Done

File: apps/livreGUI/editor/HoverPoints.h (hunk 20 above)
Comment (chevtche) [inline apps/livreGUI/editor/HoverPoints.h]: Done

File: apps/livreGUI/editor/HoverPoints.h (hunk 21 above)
Comment (chevtche) [inline apps/livreGUI/editor/HoverPoints.h]: Done

**Comment by chevtche** [comment]
restest this please

**Pull request merged**